### PR TITLE
Enable multiple input txts

### PR DESCRIPTION
**Pull request by nmstoker**
Further to discussion [here](https://discourse.mozilla.org/t/discuss-potential-pr-related-to-generate-lm-py/75883?u=nmstoker), this is my PR for enabling **generate_lm.py** to accept multiple input texts which are combined into a single lm.binary output for onward creation of a scorer.

Parameters remain unchanged so shouldn't impact current code, but if you wish to use multiple input texts, you simply pass in multiple --input_txt parameters, like so:

`python generate_lm.py --input_txt input_text_src1.txt.gz --input_txt input_text_src2.txt.gz --output_dir . --top_k 10000 --top_k 20000 --kenlm_bins path/to/kenlm/build/bin/   --arpa_order 5 --max_arpa_memory "85%" --arpa_prune "0|0|1"   --binary_a_bits 255 --binary_q_bits 8 --binary_type trie --discount_fallback`

As per the above example, you can also have corresponding top_k parameters for each input_txt. If you provider fewer top_k parameters (eg just one) then the last one will be re-used for each subsequent input_txt.

BTW: because of using repeating parameters, I didn't need the delimited for inputs I'd mentioned in the Discourse post, and this keeps it pretty simple overall.

As well as the above, I also added a simple parameter check step that runs initially (_simply to avoid situation where you have a trivial error in parameters and have to wait for the first processing step on a large text file to complete before you realise your mistake and have to run it all again!_)

And I switched it to use f-strings over .format (hope that's okay?)

**Documentation:** If this is accepted, provided people agree I plan to put a few extra basic details relating to this in the docs (eg [here](https://github.com/mozilla/DeepSpeech/blob/master/doc/Scorer.rst)) so I'd get a corresponding PR ready for that shortly too. And more generally, I had some ideas I was going to share w/ @KathyReid for the Playbook relating to this.

### CODE REVIEW

File: data/lm/generate_lm.py [hunk 1]
```diff
@@ -4,74 +4,97 @@
...
 def convert_and_filter_topk(args):
     """ Convert to lowercase, count word occurrences and save top-k words to a file """
 
-    counter = Counter()
-    data_lower = os.path.join(args.output_dir, "lower.txt.gz")
+    # Loop over multiple input_txt arguments
+    data_lowers = []
+    vocabs = []
+    vocab_str_combined = ""
+    fill = args.top_k[-1] # with multiple input_txt arguments where there are fewer top_k
```
Comment (lissyx) [inline data/lm/generate_lm.py:20]: I feel it's a hard to read design for users of the code, and likely to be a can of worm to deal with, leading to weird behaviors

File: data/lm/generate_lm.py [hunk 2]
```diff
@@ -133,23 +156,31 @@ def main():
         description="Generate lm.binary and top-k vocab for DeepSpeech."
     )
     parser.add_argument(
-        "--input_txt",
-        help="Path to a file.txt or file.txt.gz with sample sentences",
+        "--input_txt", "-i",
```
Comment (lissyx) [inline data/lm/generate_lm.py:159]: why not instead having a group options with pairs of `input_txt` and `top_k` ? As much as I remember, python argparse allows for that: https://pymotw.com/3/argparse/#argument-groups

File: data/lm/generate_lm.py [hunk 3]
```diff
@@ -133,23 +156,31 @@ def main():
...
+        help="Path to a file.txt or file.txt.gz with sample sentences. "
+             "Pass argument multiple times for multiple files.",
         type=str,
+        action='append',
         required=True,
     )
     parser.add_argument(
-        "--output_dir", help="Directory path for the output", type=str, required=True
+        "--output_dir", "-o",
```
Comment (lissyx) [inline data/lm/generate_lm.py:167]: can you avoid reformatting in the PR, or at least make that in a different commit? it's making the PR harder to follow

File: data/lm/generate_lm.py [hunk 4]
```diff
@@ -103,7 +126,7 @@ def build_lm(args, data_lower, vocab_str):
         [
             os.path.join(args.kenlm_bins, "filter"),
             "single",
-            "model:{}".format(lm_path),
+            f"model:{lm_path}",
```
Comment (lissyx) [inline data/lm/generate_lm.py:129]: This might break on Python 3.5 I think

File: data/lm/generate_lm.py (hunk 3 above)
Comment (nmstoker) [reply data/lm/generate_lm.py:167]: Yes, sorry,  I was thinking it might not have been wise to do both at once. I'll revert that side of things and leave it for later (if at all)

Comment (nmstoker) [reply data/lm/generate_lm.py:167]: Some of the reformatting was as I was trying to keep to guidance of 100 chars but I'll reduce the changes to make it easier

Comment (lissyx) [reply data/lm/generate_lm.py:167]: I'm not against the reformatting, it's just that having it in a separate commit makes it easier to review actual changes from reformatting :)

File: data/lm/generate_lm.py [hunk 5]
```diff
@@ -191,17 +222,37 @@ def main():
...
-        help="To try when such message is returned by kenlm: 'Could not calculate Kneser-Ney discounts [...] rerun with --discount_fallback'",
+        help="To try when such message is returned by kenlm: 'Could not calculate Kneser-Ney "
+             "discounts [...] rerun with --discount_fallback'",
         action="store_true",
     )
 
     args = parser.parse_args()
 
+    # Basic checks on user supplied arguments
+    if args.kenlm_bins == "path/to/kenlm/build/bin/":
```
Comment (lissyx) [inline data/lm/generate_lm.py:233]: Interesting, I would never have thought about that

File: data/lm/generate_lm.py [hunk 6]
```diff
@@ -191,17 +222,37 @@ def main():
...
+        parser.error("Update kenlm_bins from documentation value to actual path for kenlm")
+
+    if not os.path.isdir(args.kenlm_bins):
+        parser.error("kenlm_bins must be a valid directory")
+
+    kenlm_files = ["lmplz","filter","build_binary"]
```
Comment (lissyx) [inline data/lm/generate_lm.py:239]: nit: `space` after `comma` please

File: data/lm/generate_lm.py [hunk 7]
```diff
@@ -191,17 +222,37 @@ def main():
...
+    for f in kenlm_files:
+        if not os.path.isfile(os.path.join(args.kenlm_bins, f)):
+            parser.error(f"Required kenlm file {f} not found in kenlm_bins directory")
```
Comment (lissyx) [inline data/lm/generate_lm.py:242]: I think `f""` will fail on Python 3.5

File: data/lm/generate_lm.py [hunk 8]
```diff
@@ -191,17 +222,37 @@ def main():
...
+    # Basic checks on user supplied arguments
+    if args.kenlm_bins == "path/to/kenlm/build/bin/":
+        parser.error("Update kenlm_bins from documentation value to actual path for kenlm")
+
+    if not os.path.isdir(args.kenlm_bins):
+        parser.error("kenlm_bins must be a valid directory")
+
+    kenlm_files = ["lmplz","filter","build_binary"]
+    for f in kenlm_files:
+        if not os.path.isfile(os.path.join(args.kenlm_bins, f)):
```
Comment (lissyx) [inline data/lm/generate_lm.py:241]: can we also check it's executable?

File: data/lm/generate_lm.py [hunk 9]
```diff
@@ -191,17 +222,37 @@ def main():
...
+
+    if not os.path.isdir(args.output_dir):
+        parser.error("output_dir must be a valid directory")
+
     data_lower, vocab_str = convert_and_filter_topk(args)
     build_lm(args, data_lower, vocab_str)
 
     # Delete intermediate files
-    os.remove(os.path.join(args.output_dir, "lower.txt.gz"))
+    os.remove(data_lower)
```
Comment (lissyx) [inline data/lm/generate_lm.py:255]: why do we need this?

File: data/lm/generate_lm.py [hunk 10]
```diff
@@ -4,74 +4,97 @@
...
-        '  The least common word in your top-k is "{}" with {} times'.format(
-            last_word, last_count
-        )
-    )
-    for i, (w, c) in enumerate(reversed(top_counter)):
-        if c > last_count:
-            print(
-                '  The first word with {} occurrences is "{}" at place {}'.format(
-                    c, w, len(top_counter) - 1 - i
+        print(f"\nProcessing input_txt {input_txt_num} of {len(args.input_txt)}")
```
Comment (lissyx) [inline data/lm/generate_lm.py:25]: same, py3.65 and `f""`

File: data/lm/generate_lm.py [hunk 11]
```diff
@@ -4,74 +4,97 @@
...
+            else:
+                file_in = open(input_txt_item, encoding="utf-8")
+
+            for line in progressbar.progressbar(file_in):
+                line_lower = line.lower()
+                counter.update(line_lower.split())
+                file_out.write(line_lower)
+
+            file_in.close()
+
```
Comment (lissyx) [inline data/lm/generate_lm.py:51]: All of that combines reformatting and `f""` as well as indentation, would be easier if we split those apart, it makes it really hard to see the new / changed code here

File: data/lm/generate_lm.py (hunk 9 above)
Comment (nmstoker) [reply data/lm/generate_lm.py:255]: Was aiming for DRY (that file is defined previously, so seems safer to use the original definition again rather than restate a second time as was being done)

File: data/lm/generate_lm.py [hunk 12]
```diff
@@ -4,74 +4,97 @@
...
+        # Save top-k words
+        print(f"\nSaving top {top_k_item} words ...")
+        top_counter = counter.most_common(top_k_item)
+        vocab_str = "\n".join(word for word, count in top_counter)
+        vocab_str_combined = vocab_str_combined + '\n' + vocab_str
```
Comment (lissyx) [inline data/lm/generate_lm.py:56]: this is not something we had, why do we need this now?

File: data/lm/generate_lm.py [hunk 13]
```diff
@@ -4,74 +4,97 @@
...
+        vocab_path = f"vocab_{input_txt_num}-{top_k_item}.txt"
+        vocab_path = os.path.join(args.output_dir, vocab_path)
+        vocabs.append(vocab_path)
+        with open(vocab_path, "w+") as file:
+            file.write(vocab_str)
+
+        print(f"\nCalculating word statistics for input_txt {input_txt_num}...")
```
Comment (lissyx) [inline data/lm/generate_lm.py:63]: Might be worth refactoring those into a different function

File: data/lm/generate_lm.py (hunk 2 above)
Comment (nmstoker) [reply data/lm/generate_lm.py:159]: Okay, I'll look into this. So the idea is that we pair up an input_txt and a top_k for however many times the user needs them?

File: data/lm/generate_lm.py [hunk 14]
```diff
@@ -4,74 +4,97 @@
...
 
 import progressbar
 
 
 def convert_and_filter_topk(args):
     """ Convert to lowercase, count word occurrences and save top-k words to a file """
 
-    counter = Counter()
-    data_lower = os.path.join(args.output_dir, "lower.txt.gz")
+    # Loop over multiple input_txt arguments
```
Comment (lissyx) [inline data/lm/generate_lm.py:16]: To the best of my understanding, there does noi seems to be a huge difference within the body of the loop compared to the previous version, so I'm wondering if for readability we could keep `convert_and_filter_topk` a one-file level function that acts on one text file, and have a higher level function that  gets in charge of dumping everything.

The only drawback I can see for that is memory usage, so I'm wondering if maybe it is something you already experimented and came back from?

File: data/lm/generate_lm.py [hunk 15]
```diff
@@ -191,17 +222,37 @@ def main():
...
+            parser.error(f"Required kenlm file {f} not found in kenlm_bins directory")
+
+    if len(args.top_k) > len(args.input_txt):
+        parser.error("Number of top_k arguments passed should not exceed number of input_txt"
+                     " arguments passed")
+
+    if not os.path.isdir(args.output_dir):
+        parser.error("output_dir must be a valid directory")
+
     data_lower, vocab_str = convert_and_filter_topk(args)
```
Comment (lissyx) [inline data/lm/generate_lm.py:251]: It's not really your fault, but this is making me wonder whether we are not memory hungry for nothing here, with `vocab_str` and maybe we should directly write to a file that will be passed to `build_lm` instead?

File: data/lm/generate_lm.py (hunk 4 above)
Comment (nmstoker) [reply data/lm/generate_lm.py:129]: Happy to revert these `.format` changes if you recommend. I was just thinking it's easier to read plus 3.5 now seems to be EOL:
https://www.python.org/downloads/release/python-3510/

File: data/lm/generate_lm.py (hunk 6 above)
Comment (nmstoker) [reply data/lm/generate_lm.py:239]: Sorry, my bad! I did run it through cardboard but I missed that. Thanks!

File: data/lm/generate_lm.py (hunk 9 above)
Comment (lissyx) [reply data/lm/generate_lm.py:255]: Fair enough!

File: data/lm/generate_lm.py (hunk 4 above)
Comment (lissyx) [reply data/lm/generate_lm.py:129]: It is, but for now we still have CI running on it. This will change when we move to TensorFlow r2.4, but for now I feel it's an easy change.

File: data/lm/generate_lm.py (hunk 2 above)
Comment (lissyx) [reply data/lm/generate_lm.py:159]: Ineed, and so we delegate checking consistency to ArgParse instead of us :)

Comment (lissyx) [reply data/lm/generate_lm.py:159]: (and it makes cleaner for the user)

File: data/lm/generate_lm.py (hunk 12 above)
Comment (nmstoker) [reply data/lm/generate_lm.py:56]: I may have missed an alternative way to do this, but my understanding is that we need a single vocab file for passing to generate_scorer_package

In the previous "single file" approach we just passed that one vocab file over but now we've got two vocab files (one per input). Maybe I can think of a way to still get the stats on a per input basis (and broken out as you suggest below) yet in a way that it only makes a single vocab file and bypasses the need to merge them.

Does that sound better?

File: data/lm/generate_lm.py (hunk 15 above)
Comment (nmstoker) [reply data/lm/generate_lm.py:251]: I agree it probably could be made more memory efficient rather than holding the vocab string in memory all at once. I did casually look at my memory use when running but it didn't cause me issues -the machine has 64Gb RAM - but making it so it can still run over a large corpus on a lower resource machine seems wise. Will see what I can do here too.

File: data/lm/generate_lm.py (hunk 12 above)
Comment (lissyx) [reply data/lm/generate_lm.py:56]: Right, this is where the mixing of reformatting and changes made me miss the point. It makes sense now, but I still think we could benefit from splitting the loop as I explained in another comment

File: data/lm/generate_lm.py (hunk 15 above)
Comment (lissyx) [reply data/lm/generate_lm.py:251]: Yep, I have the same isssue, my working desktop has 64 CPUs, 128 threads and 128GB of RAM.

**Comment by lissyx** [comment]
Oh yeah, we will also have to cover this usecase in CI, you might want to have a look at `taskcluster/tc-scorer-tests.sh`: https://github.com/mozilla/DeepSpeech/blob/385c8c769bc4aed5e6979a239591486c44f3471d/taskcluster/tc-scorer-tests.sh

File: data/lm/generate_lm.py (hunk 14 above)
Comment (lissyx) [reply data/lm/generate_lm.py:16]: @nmstoker I think this would be the biggest change in your PR, what's your take on that?

Comment (nmstoker) [reply data/lm/generate_lm.py:16]: Yes, I think this sounds like a good way forward.

Is it cleanest for me to do a new PR, handling the steps as distinct commits? (as I should've done first time!)

I was thinking I'd keep the keep `convert_and_filter_topk` a one-file level function, I'd try to avoid it passing the whole vocab in memory (as per the other comment), I'd split out the stats like you suggested and lastly I'd get it looping over the multiple files.

If okay, I'll get that sorted tomorrow or Sunday at the latest.

Comment (lissyx) [reply data/lm/generate_lm.py:16]: > Is it cleanest for me to do a new PR, handling the steps as distinct commits? (as I should've done first time!)

No, just rebase / edit your history and force push?



> If okay, I'll get that sorted tomorrow or Sunday at the latest.

Sounds awesome!

Comment (nmstoker) [reply data/lm/generate_lm.py:16]: > No, just rebase / edit your history and force push?

Great, I'll do that 👍🏻

**Comment by nmstoker** [comment]
Quick update: I got unexpectedly delayed with some things that came up, so I made less progress than I'd hoped but I'll continue it in the evenings this week/next weekend to get this sorted.

**Comment by lissyx** [comment]
> Quick update: I got unexpectedly delayed with some things that came up, so I made less progress than I'd hoped but I'll continue it in the evenings this week/next weekend to get this sorted.

Gentle ping?

**Comment by nmstoker** [comment]
Sorry, I haven't had a great deal of time so this isn't likely to proceed much for a bit. I'll try a bit more next weekend

**Comment by lissyx** [comment]
> Sorry, I haven't had a great deal of time so this isn't likely to proceed much for a bit. I'll try a bit more next weekend

@nmstoker Gentle ping? It's okay if you don't have time ;)
